src/main/java/com/liferay/faces/bridge/ext/atmosphere/internal/AnnotationProcessorLiferayImpl.java

```diff
@@ -0,0 +1,267 @@
+/**
+ * Copyright (c) 2000-2017 Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+package com.liferay.faces.bridge.ext.atmosphere.internal;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.servlet.annotation.HandlesTypes;
+
+import org.atmosphere.config.AtmosphereAnnotation;
+
+import org.atmosphere.cpr.AnnotationHandler;
+import org.atmosphere.cpr.AnnotationProcessor;
+import org.atmosphere.cpr.AnnotationScanningServletContainerInitializer;
+import org.atmosphere.cpr.AtmosphereConfig;
+import org.atmosphere.cpr.AtmosphereFramework;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.wiring.BundleWiring;
+
+import com.liferay.faces.util.logging.Logger;
+import com.liferay.faces.util.logging.LoggerFactory;
+
+
+/**
+ * @author  Kyle Stiemann
+ */
+public class AnnotationProcessorLiferayImpl implements AnnotationProcessor {
```

Comment:
> How does Atmosphere discover this class?

[Via **`web.xml`** `<init-param>`](https://github.com/liferay/liferay-faces-bridge-impl/blob/master/demo/primepush-cdi-counter-portlet/src/main/webapp/WEB-INF/web-liferay.xml#L70-L73):

```
<servlet>
	<servlet-name>Push Servlet</servlet-name>
	<servlet-class>org.primefaces.push.PushServlet</servlet-class>
	<!-- ...other init-params here... -->
	<init-param>
		<param-name>org.atmosphere.cpr.AnnotationProcessor</param-name>
		<param-value>com.liferay.faces.bridge.ext.atmosphere.internal.AnnotationProcessorLiferayImpl</param-value>
	</init-param>
</servlet>
```

> Also, if com.liferay.faces.bridge.ext.jar is deployed as an OSGi module, then would Atmosphere be able to discover it since it is not part of the API?

Hmm, I'm not sure. Should we just make this class API since it is used in the **`web.xml`** already?

> Also is it OK to have an <optional>true</optional> dependency for an OSGi module?

[In OSGi, you can mark package dependencies as `resolution:=optional`](http://enroute.osgi.org/tutorial_wrap/220-optional-dependencies.html). So, yes optional dependencies are fine.